File: src/main/js/formulas/math/ROUNDDOWN.ts
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ROUNDDOWN extends AbstractFormula {
  static SELF: ROUNDDOWN = new ROUNDDOWN();
  NAME = FormulaName.ROUNDDOWN;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLengthBetween(values.length, 1, 2, this.NAME);
    const value = Converters.first(this.collateralLookup(origin, values[0]));
    let placesRaw;
    if (values.length === 1) {
      placesRaw = 0.0;
    } else {
      placesRaw = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[1])));
    }
    const n = Converters.toNumber(value);
    const places = placesRaw > 0 ? Math.floor(placesRaw) : Math.ceil(placesRaw);
    if (n < 0) {
      if (places < 0) {
        return Number(Math.ceil(parseFloat(n + "e-" + places * -1)) + "e" + places * -1);
      }
      return Number(Math.ceil(parseFloat(n + "e" + places)) + "e-" + places);
    }
    if (places < 0) {
      return Number(Math.floor(parseFloat(n + "e-" + places * -1)) + "e" + places * -1);
    }
    return Number(Math.floor(parseFloat(n + "e" + places)) + "e-" + places);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/ROUNDDOWN.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.function.BiFunction;
import java.util.function.Function;

public class ROUNDDOWN extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ROUNDDOWN;
  public static ROUNDDOWN SELF = new ROUNDDOWN();

  public ROUNDDOWN() {
    super();
  }

  public ROUNDDOWN(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLengthBetween(values.length, 1, 2, NAME);
    Object value = Converters.first(collateralLookup.apply(origin, values[0]));
    Double placesRaw;
    if (values.length == 1) {
      placesRaw = 0.0;
    } else {
      placesRaw = Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[1])));
    }
    Double number = Converters.toDouble(value);
    BigDecimal bd = new BigDecimal(number.toString());
    int places;
    if (placesRaw > 0) {
      places = new Double(Math.floor(placesRaw.floatValue())).intValue();
    } else {
      places = new Double(Math.ceil(placesRaw.floatValue())).intValue();
    }
    if (number < 0) {
      bd = bd.setScale(places, RoundingMode.CEILING);
    } else {
      bd = bd.setScale(places, RoundingMode.FLOOR);
    }
    return bd.doubleValue();
  }
}
